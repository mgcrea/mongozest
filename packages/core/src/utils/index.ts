export * from './logger';
export * from './objectId';
export * from './traversing';
