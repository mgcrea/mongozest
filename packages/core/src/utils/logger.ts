import chalk from 'chalk';
import {inspect as defaultInspect} from 'util';
import {toString} from 'lodash';

export {chalk};

export const chalkString = (s: string): string => chalk.green(`'${s}'`);
export const chalkStringArray = (a: Array<string>): string => `[ ${a.map(chalkString).join(', ')} ]`;
export const chalkNumber = (n: number | string): string => chalk.yellow(toString(n));
export const chalkBoolean = (b: boolean): string => chalk.yellow(b ? 'true' : 'false');
export const chalkDate = (d: Date): string => chalk.magenta(d.toISOString());

export const inspect = (maybeObject: unknown): string =>
  defaultInspect(maybeObject, {compact: true, colors: true, depth: Infinity, breakLength: Infinity});

export const log = (maybeString: string): void =>
  console.log(`${chalk.gray(new Date().toISOString())} - 🏛 mongodb: ${maybeString}`);
