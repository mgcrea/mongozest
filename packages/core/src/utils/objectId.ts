import {intersectionWith, unionWith, uniqBy, uniqWith} from 'lodash';
import {ObjectId} from 'mongodb';

export const includesObjectId = (arrayOfObjectIds: ObjectId[], targetObjectId: ObjectId) =>
  arrayOfObjectIds.some(objectId => objectId.equals(targetObjectId));

export const intersectionWithObjectIds = (...arraysOfObjectIds: ObjectId[]) =>
  intersectionWith<ObjectId, ObjectId>(...arraysOfObjectIds, (arrVal: ObjectId, othVal: ObjectId) =>
    arrVal.equals(othVal)
  );

export const unionWithObjectIds = (...arraysOfObjectIds: ObjectId[]) =>
  unionWith<ObjectId>(...arraysOfObjectIds, (arrVal: ObjectId, othVal: ObjectId) => arrVal.equals(othVal));

export const uniqWithObjectIds = (arrayOfObjectIds: ObjectId[]) =>
  uniqWith(arrayOfObjectIds, (arrVal, othVal) => arrVal.equals(othVal));

export const uniqByObjectId = (arrayOfObjects: {_id: ObjectId}[]) =>
  uniqBy(arrayOfObjects, arrVal => arrVal._id.toString());
