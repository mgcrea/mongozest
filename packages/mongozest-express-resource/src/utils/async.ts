export const asyncHandler = (wrappedFunction: Function) =>
  function asyncWrap(...args: any[]) {
    const next = args[args.length - 1];
    return new Promise(function(resolve) {
      resolve(wrappedFunction(...args));
    }).catch(next);
  };
