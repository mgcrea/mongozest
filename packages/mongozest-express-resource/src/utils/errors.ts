import createError from 'http-errors';
import {Request, Response, NextFunction} from 'express';
import {MongoError} from '@mgcrea/mongozest';

export const mongoErrorMiddleware = (err: any, req: Request, res: Response, next: NextFunction) => {
  // Convert MongoError to HttpErrors
  if (err instanceof MongoError) {
    const {errmsg = '', code} = err;
    // Handle Validation errors
    if (err.code === 121) {
      next(createError(422, errmsg));
      return;
    }
    // Handle duplicate key errors (check me!)
    if (err.code === 110) {
      next(createError(409, errmsg));
      return;
    }
    next(createError(400, errmsg));
    return;
  }
  next(err);
};
