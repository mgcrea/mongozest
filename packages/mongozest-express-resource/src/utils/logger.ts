import chalk from 'chalk';
import {inspect as defaultInspect} from 'util';

export const inspect = (maybeObject: object) =>
  defaultInspect(maybeObject, {compact: true, colors: true, depth: Infinity, breakLength: Infinity});

export const log = (maybeString: string) =>
  console.log(`${chalk.gray(new Date().toISOString())} - 🏢 resource: ${maybeString}`);
