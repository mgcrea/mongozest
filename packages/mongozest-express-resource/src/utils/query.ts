// import JSON5 from 'json5';
// import {defaults, pick, isPlainObject, mapValues} from 'lodash';
// import createError from 'http-errors';

// export const parseQueryOptions = (req, whitelist = []) => {
//   // Only pick valid query options
//   let queryOptions = defaults(pick(req.query, whitelist), {});
//   // Parse any JSON-like options
//   queryOptions = mapValues(queryOptions, (value, key) => {
//     if (value && typeof value === 'string' && (value[0] === '{' || value[0] === '[')) {
//       try {
//         return JSON5.parse(value);
//       } catch (err) {
//         throw createError(400, `Failed to parse query field=\`${key}\``);
//       }
//     }
//     return value;
//   });
//   // Extend with req parameters
//   if (req.params && Object.keys(req.params).length) {
//     queryOptions.query = Object.assign(queryOptions.query || {}, req.params);
//   }
//   return queryOptions;
// };

// export const parseQuerySelect = req => {
//   const query = parseQueryOptions(req, ['select']);
//   if (isPlainObject(query.select)) {
//     return query.select;
//   }
//   const querySelectAsArray = !Array.isArray(query.select) ? query.select.toString().split(' ') : query.select;
//   return querySelectAsArray.reduce((soFar, value) => {
//     if (value.startsWith('+')) {
//       soFar[value.substr(1)] = 1;
//     } else if (value.startsWith('-')) {
//       soFar[value.substr(1)] = 0;
//     } else {
//       soFar[value] = 1;
//     }
//     return soFar;
//   }, {});
// };

// export const applyQueryOptions = (query, queryOptions = {}) => {
//   if (queryOptions.distinct) {
//     query.distinct(queryOptions.distinct);
//   }
//   if (queryOptions.limit > 0) {
//     query.limit(queryOptions.limit * 1);
//   }
//   if (queryOptions.select) {
//     if (Array.isArray(queryOptions.select)) {
//       query.select(queryOptions.select.join(' '));
//     } else {
//       query.select(queryOptions.select);
//     }
//   }
//   if (queryOptions.populate) {
//     query.populate(queryOptions.populate);
//   }
//   if (queryOptions.sort) {
//     query.sort(queryOptions.sort);
//   }
//   return query;
// };
