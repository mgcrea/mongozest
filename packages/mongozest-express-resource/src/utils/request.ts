export const parseBodyAsUpdate = (body: {[s: string]: any}) =>
  Object.keys(body).reduce(
    (soFar: {[s: string]: any}, key: string) => {
      if (!key.startsWith('$')) {
        soFar.$set[key] = body[key];
      } else {
        soFar[key] = body[key];
      }
      return soFar;
    },
    {$set: {}}
  );

export const asyncHandler = (wrappedFunction: Function) =>
  function asyncWrap(...args: any[]) {
    const next = args[args.length - 1];
    return new Promise(function(resolve) {
      resolve(wrappedFunction(...args));
    }).catch(err => {
      d(err);
      next(err);
    });
  };
