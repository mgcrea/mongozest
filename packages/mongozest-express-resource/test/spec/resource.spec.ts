import {makeFetch} from 'supertest-fetch';
import {ObjectId, Model} from '@mgcrea/mongozest';
import {omit, isObject} from 'lodash';
import {createTestApp, getDbName, breakdownMiddleware, fixtures} from './../utils';
import createResource, {Resource} from './../../src';

const DB_NAME = getDbName(__filename);

const app = createTestApp({routers: []});
const {mongo, redis, insertFixture} = app.locals;
app.locals.fixtures = fixtures;
const fetch = makeFetch(app);

interface UserSchema {
  firstName?: string;
  lastName?: string;
  email: string;
  nationality?: string;
  device?: ObjectId;
}

class User extends Model<UserSchema> {
  static schema = {
    firstName: {bsonType: 'string'},
    lastName: {bsonType: 'string'},
    email: {bsonType: 'string', required: true},
    nationality: {bsonType: 'string'},
    device: {bsonType: 'objectId'}
  };
}

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
  await mongo.loadModel(User);
});

afterAll(async () => {
  await Promise.all([redis.quit(), mongo.disconnect()]);
});

describe('Resource', () => {
  let resource: Resource;
  describe('resource', () => {
    it('should properly create resource', async () => {
      resource = createResource('User', {db: 'mongo'});
      expect(resource instanceof Resource).toBeTruthy();
    });
    it('should properly serve resource', async () => {
      const router = resource.build();
      app.use(router);
      app.use(breakdownMiddleware);
    });
  });

  describe('collection', () => {
    describe('GET /users', () => {
      it('should return 200', async () => {
        const {insertedId} = await insertFixture('User.mongozest');
        const res = await fetch(`/users`, {
          method: 'get',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Array.isArray(resBody)).toBeTruthy();
        expect(resBody.length).toEqual(1);
        expect(omit(resBody[0], '_id')).toMatchSnapshot();
      });
    });
    describe('POST /users', () => {
      it('should return 200 when payload is valid', async () => {
        const reqBody = {
          firstName: 'John',
          lastName: 'Doe',
          email: 'john.doe@gmail.com'
        };
        const res = await fetch('/users', {
          method: 'post',
          body: JSON.stringify(reqBody),
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Object.keys(resBody)).toMatchObject([...Object.keys(reqBody), '_id']);
        expect(ObjectId.isValid(resBody._id)).toBeTruthy();
      });
      it('should return 400 when payload is invalid', async () => {
        const reqBody = {
          firstName: 'John',
          lastName: 'Doe'
        };
        const res = await fetch('/users', {
          method: 'post',
          body: JSON.stringify(reqBody),
          headers: {'Content-Type': 'application/json'}
        })
          .expect(422)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Object.keys(resBody)).toMatchObject(['error']);
        expect(resBody).toMatchSnapshot();
      });
    });
    describe('PATCH /users', () => {
      it('should return 200', async () => {
        const {insertedId} = await insertFixture('User.mongozest');
        const reqBody = {
          firstName: 'Alex'
        };
        const res = await fetch(`/users`, {
          method: 'patch',
          body: JSON.stringify(reqBody),
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Array.isArray(resBody)).toBeTruthy();
        expect(resBody.length > 0).toBeTruthy();
        expect(omit(resBody[0], '_id')).toMatchSnapshot();
      });
    });
    describe('DELETE /users', () => {
      it('should return 200', async () => {
        await insertFixture('User.mongozest');
        const res = await fetch(`/users`, {
          method: 'delete',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
      });
    });
  });

  describe('document', () => {
    describe('GET /users/:_id', () => {
      it('should return 200', async () => {
        const {_id: insertedId} = await insertFixture('User.mongozest');
        const res = await fetch(`/users/${insertedId}`, {
          method: 'get',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(omit(resBody, '_id')).toMatchSnapshot();
      });
    });
    describe('PATCH /users/:_id', () => {
      it('should return 200', async () => {
        const {_id: insertedId} = await insertFixture('User.mongozest');
        const reqBody = {
          firstName: 'Laura'
        };
        const res = await fetch(`/users/${insertedId}`, {
          method: 'patch',
          body: JSON.stringify(reqBody),
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(omit(resBody, '_id')).toMatchSnapshot();
      });
    });
    describe('DELETE /users/:_id', () => {
      it('should return 200', async () => {
        const {_id: insertedId} = await insertFixture('User.mongozest');
        const res = await fetch(`/users/${insertedId}`, {
          method: 'delete',
          headers: {'Content-Type': 'application/json'}
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(isObject(resBody)).toBeTruthy();
        expect(Object.keys(resBody)).toMatchSnapshot();
        expect(resBody).toMatchSnapshot();
      });
    });
  });
});

describe('resource with nested paths', () => {
  let resource: Resource;
  const PATH: string = '/devices/:device/users';
  it('should properly create resource', async () => {
    resource = createResource('User', {
      db: 'mongo',
      paths: [PATH],
      params: {
        device: (_id: string) => {
          return {device: ObjectId.createFromHexString(_id)};
        }
      }
    });
    expect(resource instanceof Resource).toBeTruthy();
  });
  it('should properly serve resource', async () => {
    const router = resource.build();
    app.use(router);
    app.use(breakdownMiddleware);
  });
  // describe('collection', () => {
  //   describe('GET /users', () => {
  //     it('should return 200', async () => {
  //       const {insertedId} = await insertFixture('User.mongozest');
  //       const res = await fetch(`/users`, {
  //         method: 'get',
  //         headers: {'Content-Type': 'application/json'}
  //       })
  //         .expect(200)
  //         .expect('content-type', /^application\/json/);
  //       const resBody = await res.json();
  //       expect(Array.isArray(resBody)).toBeTruthy();
  //       expect(resBody.length).toEqual(1);
  //       expect(omit(resBody[0], '_id')).toMatchSnapshot();
  //     });
  //   });
  //   describe('POST /users', () => {
  //     it('should return 200 when payload is valid', async () => {
  //       const reqBody = {
  //         firstName: 'John',
  //         lastName: 'Doe',
  //         email: 'john.doe@gmail.com'
  //       };
  //       const res = await fetch('/users', {
  //         method: 'post',
  //         body: JSON.stringify(reqBody),
  //         headers: {'Content-Type': 'application/json'}
  //       })
  //         .expect(200)
  //         .expect('content-type', /^application\/json/);
  //       const resBody = await res.json();
  //       expect(Object.keys(resBody)).toMatchObject([...Object.keys(reqBody), '_id']);
  //       expect(ObjectId.isValid(resBody._id)).toBeTruthy();
  //     });
  //     it('should return 400 when payload is invalid', async () => {
  //       const reqBody = {
  //         firstName: 'John',
  //         lastName: 'Doe'
  //       };
  //       const res = await fetch('/users', {
  //         method: 'post',
  //         body: JSON.stringify(reqBody),
  //         headers: {'Content-Type': 'application/json'}
  //       })
  //         .expect(422)
  //         .expect('content-type', /^application\/json/);
  //       const resBody = await res.json();
  //       expect(Object.keys(resBody)).toMatchObject(['error']);
  //       expect(resBody).toMatchSnapshot();
  //     });
  //   });
  // });
});
