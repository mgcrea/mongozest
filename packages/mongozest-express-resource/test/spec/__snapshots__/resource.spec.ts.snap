// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource collection GET /users should return 200 1`] = `
Object {
  "email": "john.doe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;

exports[`Resource collection POST /users should return 400 when payload is invalid 1`] = `
Object {
  "error": "Document failed validation",
}
`;

exports[`Resource collection document GET /users/:id should return 200 1`] = `
Object {
  "email": "john.doe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;
